tests/unit/cunumeric/test_utils.py

```diff
@@ -0,0 +1,250 @@
+        result = m.find_last_user_frames(top_only=False)
+        assert isinstance(result, str)
+        assert "|" in result
+
+        # it would be too fragile to compare more specific than this
+        assert len(result.split("|")) > 1
+        assert all(len(x.split(":")) == 2 for x in result.split("|"))
+
+
+class Test_is_supported_dtype:
+    @pytest.mark.parametrize(
+        "value", ["foo", 10, 10.2, [], (), {}, set(), None]
+    )
+    def test_type_bad(self, value) -> None:
+        with pytest.raises(AssertionError):  # should be TypeError?
+            m.is_supported_dtype(value)
+
+    @pytest.mark.parametrize("value", SUPPORTED_DTYPES)
+    def test_supported(self, value) -> None:
+        assert m.is_supported_dtype(np.dtype(value))
+
+    # This is just a representative sample, not exhasutive
+    @pytest.mark.parametrize(
+        "value", [np.float128, np.complex64, np.datetime64]
+    )
+    def test_unsupported(self, value) -> None:
+        assert not m.is_supported_dtype(np.dtype(value))
+
+
+@pytest.mark.parametrize(
+    "shape, volume", [[(), 0], [(10,), 10], [(1, 2, 3), 6]]
+)
+def test_calculate_volume(shape, volume) -> None:
+    assert m.calculate_volume(shape) == volume
+
+
+def test_get_arg_dtype() -> None:
+    dt = m.get_arg_dtype(np.float32)
+    assert dt.type is np.void
+    assert dt.isalignedstruct
+    assert set(dt.fields) == {"arg", "arg_value"}
+    assert dt.fields["arg"][0] == np.dtype(np.int64)
+    assert dt.fields["arg_value"][0] == np.dtype(np.float32)
+
+
+def test_get_arg_value_dtype() -> None:
+    dt = m.get_arg_dtype(np.float32)
+    assert m.get_arg_value_dtype(dt) is np.float32
+
+
+def test_dot_modes() -> None:
+    assert m.dot_modes(0, 0) == ([], [], [])
+    assert m.dot_modes(0, 1) == ([], ["A"], ["A"])
+    assert m.dot_modes(1, 0) == (["a"], [], ["a"])
+    assert m.dot_modes(1, 1) == (["a"], ["a"], [])
+    assert m.dot_modes(2, 0) == (["a", "b"], [], ["a", "b"])
+    assert m.dot_modes(0, 2) == ([], ["A", "B"], ["A", "B"])
+    assert m.dot_modes(2, 1) == (["a", "b"], ["b"], ["a"])
+    assert m.dot_modes(1, 2) == (["a"], ["a", "B"], ["B"])
+    assert m.dot_modes(2, 2) == (["a", "b"], ["b", "B"], ["a", "B"])
+
+
+def test_inner_modes() -> None:
+    assert m.inner_modes(0, 0) == ([], [], [])
+    assert m.inner_modes(0, 1) == ([], ["A"], ["A"])
+    assert m.inner_modes(1, 0) == (["a"], [], ["a"])
+    assert m.inner_modes(1, 1) == (["a"], ["a"], [])
+    assert m.inner_modes(2, 0) == (["a", "b"], [], ["a", "b"])
+    assert m.inner_modes(0, 2) == ([], ["A", "B"], ["A", "B"])
+    assert m.inner_modes(2, 1) == (["a", "b"], ["b"], ["a"])
+    assert m.inner_modes(1, 2) == (["a"], ["A", "a"], ["A"])
+    assert m.inner_modes(2, 2) == (["a", "b"], ["A", "b"], ["a", "A"])
+
+
+@pytest.mark.parametrize("a_ndim, b_ndim", [(0, 0), (0, 1), (1, 0)])
+def test_matmul_modes_bad(a_ndim, b_ndim) -> None:
+    with pytest.raises(AssertionError):  # should be ValueError?
```

Comment:
These functions are for internal use, so it should probably be the `InternalError` that you had suggested previously.

Reply:
But `ValueError` is fine too, please change to whatever you think is more appropriate.

Reply:
Now that I look at this code again, the reason why `tensordot_modes` has so much error checking is because the values are fed straight from user input, while `matmul_modes` gets some minimal sanitization before being called. It's probably more consistent to just move this check inside `matmul_modes`: https://github.com/nv-legate/cunumeric/blob/branch-22.05/cunumeric/module.py#L2458-L2459

Reply:
@manopapad added here https://github.com/nv-legate/cunumeric/pull/229/commits/1c6d0e94bdd1a9ed1fe9637efed09903cfedaac7 

note I also changed another assert to raise a `TypeError` please let me know if that seems OK

Reply:
lgtm